### PR TITLE
Update to use new ASDL implementation.

### PR DESCRIPTION
**Pull request by alexreinking**
* Consolidate (most) grammars in grammars.py
* Fix immutability violations caught by test cases.
* Fix multiple-path imports. (Imports are still a mess)
* Remove ASDL as submodule, use the PyPI-published copy.
---
Upcoming (not for this PR):
* Generate `.pyi` from ASDL (hence the consolidation)

### CODE REVIEW

**Comment by yamaguchi1024** [comment]
What was the reason behind moving LoopIR, UAST, Effects, and PAST to a single file?

**Comment by alexreinking** [comment]
> What was the reason behind moving LoopIR, UAST, Effects, and PAST to a single file?

So they can share validator types and refer to each other without requiring circular imports.

**Comment by yamaguchi1024** [comment]
Ah, nice!!

**Review by yamaguchi1024 — APPROVED** [review]

File: SYS_ATL/LoopIR_unification.py [hunk 1]
```diff
@@ -612,7 +612,7 @@ def __init__(self, subproc, stmt_block, live_vars):
 
         self.stride_holes   = { fa.name : False for fa in subproc.args
                                                 if fa.type == T.stride }
-        
+
```
Comment (alexreinking) [inline SYS_ATL/LoopIR_unification.py:615]: If you want small diffs, you'll have to use automatic source formatting (or at _least_ whitespace normalization) tools...

File: SYS_ATL/new_eff.py [hunk 2]
```diff
@@ -71,7 +62,7 @@ def __add__(lhs, rhs):
                 rc = rhs.coords[ri]
                 ri += 1
                 coords.append( AWinCoord(rc.is_pt, rc.val + lc.val) )
-        return AWin(lhs.name, coords, lhs.strides) 
+        return AWin(lhs.name, coords, lhs.strides)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:65]: More stupid whitespace diffs. I'm not reverting these.

File: SYS_ATL/new_eff.py [hunk 3]
```diff
@@ -669,7 +665,7 @@ def is_empty(ls):
 
 
 E = ADT("""
-module Effects {
+module EffectsNew {
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:668]: The module name determines the entry in `sys.modules` so name clashes are incorrect. Detecting this error is TODO in asdl-adt.

File: SYS_ATL/new_eff.py [hunk 4]
```diff
@@ -1339,7 +1335,7 @@ def Check_DeleteConfigWrite(proc, stmts):
     # the statement block being focused on.  Filter out any
     # such configuration variables whose values are definitely unchanged
     def is_cfg_unmod_by_stmts(pt):
-        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1338]: _Whitespace normalization_

File: SYS_ATL/new_eff.py [hunk 5]
```diff
@@ -1349,7 +1345,7 @@ def is_cfg_unmod_by_stmts(pt):
     # consider every global that might be modified
     cfg_mod_visible = set()
     for _,pt in cfg_mod.items():
-        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1348]: _Whitespace normalization_

File: SYS_ATL/new_eff.py [hunk 6]
```diff
@@ -1532,5 +1528,3 @@ def Check_Bounds(proc, alloc_stmt, block):
     if not is_ok:
         raise SchedulingError(
             f"The buffer {alloc_stmt.name} is accessed out-of-bounds")
-
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1535]: _Whitespace normalization_

File: SYS_ATL/prec_analysis.py [hunk 7]
```diff
@@ -101,19 +101,20 @@ def map_s(self, s):
             assert ltyp != T.err and ltyp != T.R
 
             # update the type annotation here if needed
-            result[0].type = ltyp
+            result[0] = result[0].update(type=ltyp)
```
Comment (alexreinking) [inline SYS_ATL/prec_analysis.py:104]: ADT nodes are now immutable (and this is checked at runtime). Use `.update()` to create a new node with a subset of fields changed.

File: SYS_ATL/prec_analysis.py [hunk 8]
```diff
@@ -101,19 +101,20 @@ def map_s(self, s):
...
-                    result[0].rhs = self.coerce_e(result[0].rhs, ltyp)
+                    result[0] = result[0].update(
+                        rhs=self.coerce_e(result[0].rhs, ltyp))
                     rtyp = ltyp
 
                 # TODO: remove the `cast` field entirely
                 if ltyp != rtyp:
                     # then we have an implicit cast at this point
-                    result[0].cast = "yup, cast!"
+                    result[0] = result[0].update(cast="yup, cast!")
```
Comment (alexreinking) [inline SYS_ATL/prec_analysis.py:117]: This just _screams_ "bug". How did this even get checked in?

File: SYS_ATL/LoopIR.py [hunk 9]
```diff
@@ -1,34 +1,132 @@
...
+            return super().__new__(cls, op)
+        raise ValueError(f'invalid operator: {op}')
+
+
+# --------------------------------------------------------------------------- #
+# Loop IR
+# --------------------------------------------------------------------------- #
+
+
+LoopIR = ADT("""
```
Comment (alexreinking) [inline SYS_ATL/LoopIR.py:53]: Moving this up to allow the later ADTs to reference `LoopIR.type`

**Comment by alexreinking** [comment]
I'm not worried about a 0.01% decrease in coverage. That will go up when we inevitably add more tests later.

**Pull request merged**